### PR TITLE
[DOCS] Retrievers and rerankers

### PR DESCRIPTION
**Pull request by leemthompo**
## Background 
Semantic reranking docs require 3 components:
1. The bare minimum API docs stuff in the [Search API / Retriever](https://www.elastic.co/guide/en/elasticsearch/reference/current/retriever.html) API reference (normally devs make these updates)
2. I'll need to update the high-level [retrievers overview](https://www.elastic.co/guide/en/elasticsearch/reference/current/retrievers-overview.html) that lives in "Search your data" docs section
3. New semantic reranking overview page, sibling to retrievers overview under new *Retrievers and reranking* section

## Summary 
- This PR focuses on number 3, new high-level information about semantic reranking, with some small updates to 2.
- [EDIT] Drafted 1 also in https://github.com/elastic/elasticsearch/pull/110007/commits/499757723f4b86d799c6cc6c030f5b6a29d652e7

### ✨ [URL preview](https://elasticsearch_bk_110007.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/semantic-reranking.html)

### CODE REVIEW

**Comment by leemthompo** [comment]
run docs-build

**Comment by leemthompo** [comment]
@elasticsearchmachine test this

**Comment by leemthompo** [comment]
@elasticmachine update branch

File: docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+====
+
+Rerankers improve the relevance of results from earlier-stage retrieval mechanisms.
+_Semantic_ rerankers use machine learning models to reorder search results based on their semantic similarity to a query.
+
+First-stage retrievers and rankers must be very fast and efficient because they process either the entire corpus, or all matching documents.
+In a multi-stage pipeline, you can progressively use more computationally intensive ranking functions and techniques, as they will operate on smaller result sets at each step.
+This helps avoid query latency degradation and keeps costs manageable.
+
+Semantic reranking requires relatively large and complex machine learning models and operates in real-time in response to queries.
```
Comment (joshdevins) [inline docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc:18]: Maybe here extend this to add "and that's why we use it only on a small top-k, as one the of the last steps in a pipeline". Don't quote me on that 😆

File: docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+Semantic reranking enables a variety of use cases:
+
+* *Lexical (BM25) retrieveal results reranking*
+** Out-of-the-box semantic search by adding a simple API call to any lexical/BM25 retrieval pipeline.
+** Add semantic search capabilities on top of existing indices without reindexing, perfect for quick improvements.
+** Ideal for environments with complex existing indices.
+
+* *Semantic retrieval results reranking*
+** Improves results from semantic retrievers using ELSER sparse vector embeddings or dense vector embeddings by using more powerful models.
+
```
Comment (joshdevins) [inline docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc:37]: Maybe add a second bullet like, "Adds a refinement layer ontop of hybrid retrieval with RRF"

File: docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+* *General applications*
+** Supports automatic and transparent chunking, eliminating the need for pre-chunking at index time.
+** Provides explicit control over document relevance in retrieval-augmented generation (RAG) uses cases or other scenarios involving language model (LLM) inputs.
+
+Now that we've outlined the value of semantic reranking, we'll explore the specific models that power this process and how they differ.
+
+[discrete]
+[[semantic-reranking-models]]
+==== Cross-encoder and bi-encoder models for semantic reranking
```
Comment (joshdevins) [inline docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc:46]: We only support cross-encoders today, so this section can just focus on those models.

Comment (joshdevins) [reply docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc:46]: Save this for future iterations 😉

File: docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+[discrete]
+[[semantic-reranking-in-es]]
+==== Semantic reranking in {es}
+
+In {es}, semantic rerankers are implemented using the {es} *Inference API* and a <<retriever,retriever>>.
+
+To use semantic reranking in {es}, you need to:
+
+. Choose a reranking model. In addition to cross-encoder and bi-encoder models running on {es} inference nodes, we also expose external models and services via the Inference API to semantic rerankers.
```
Comment (joshdevins) [inline docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc:87]: Remove reference to bi-encoder rerankers

**Review by joshdevins — DISMISSED** [review]
Generally looks fine, but since we don't support bi-encoders, we need to remove those references and rejig the content at the end a bit.

File: docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      },
+      "field": "text",
+      "inference_id": "my-cohere-rerank-model-v3",
+      "inference_text": "How often does the moon hide the sun?",
+      "rank_window_size": 100,
+      "min_score": 0.5
+    }
+  },
+  "_source": ["title", "text"],
+  "size": 10
```
Comment (demjened) [inline docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc:126]: These properties are not related to reranking, I suggest removing them in order to simplify the example.
```suggestion
  }
```

**Review by demjened — CHANGES_REQUESTED** [review]
@leemthompo Looks great - I've left some comments and filled in a couple of sections, please review them.

**Comment by leemthompo** [comment]
Thanks @demjened for the keen-eyed review 👍

File: docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+For example, this is important when using results in a RAG workflow or if you're otherwise feeding results to LLMs.
+Note that similarity scores from bi-encoders/embedding similarities are _query-dependent_, meaning you cannot set universal cut-offs.
+* Bi-encoders rerank using embeddings. You can improve your reranking latency by creating embeddings at ingest-time. These embeddings can be stored for reranking without being indexed for retrieval, reducing your memory footprint.
+==============
+
+[discrete]
+[[semantic-reranking-in-es]]
+==== Semantic reranking in {es}
+
+In {es}, semantic rerankers are implemented using the {es} <<inference-apis,*Inference API*>> and a <<retriever,retriever>>.
```
Comment (demjened) [inline docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc:88]: Nit: bolding is not required.
```suggestion
In {es}, semantic rerankers are implemented using the {es} <<inference-apis,Inference API>> and a <<retriever,retriever>>.
```

File: docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        "standard": {
+          "query": {
+            "match": {
+              "text": "How often does the moon hide the sun?"
+            }
+          }
+        }
+      },
+      "field": "text",
+      "inference_id": "my-cohere-rerank-model-v3",
```
Comment (demjened) [inline docs/reference/search/search-your-data/retrievers-reranking/semantic-reranking.asciidoc:119]:
```suggestion
      "inference_id": "my-cohere-rerank-model",
```

File: docs/reference/search/retriever.asciidoc [hunk 8]
```diff
@@ -201,6 +204,70 @@ GET /index/_search
...
+GET /index/_search
+{
+    "retriever": {
+        "text_similarity_reranker": {
+            "retriever": {
+                "standard": { ... }
+            }
+        },
+        "field": "text",
+        "inference_id": "my-cohere-rerank-model-v3",
```
Comment (demjened) [inline docs/reference/search/retriever.asciidoc:262]: Nit: sorry, I left the `-v3` in due to a copy-paste because I used the Cohere Reranker English v3 model. To the reader it could be distracting, so let's remove it.

```suggestion
        "inference_id": "my-cohere-rerank-model",
```

**Review by demjened — APPROVED** [review]
LGTM, just a couple of nitpicks.
